scylla-cdc/src/reader.rs

```diff
@@ -0,0 +1,298 @@
+                println!("pk: {} t: {} v: {} s: {}", pk, t, v, s);
+            }
+
+            if lower_timestamp == upper_timestamp {
+                break;
+            }
+
+            lower_timestamp = upper_timestamp;
+            sleep(self.sleep_interval).await;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::error::Error;
+
+    use scylla::query::Query;
+    use scylla::{SessionBuilder};
+    use scylla::statement::Consistency;
+
+    use super::*;
+
+    const TEST_KEYSPACE: &str = "test";
+    const TEST_TABLE: &str = "t";
+    const SLEEP_INTERVAL: u64 = 1;
+    const WINDOW_SIZE: i64 = 2;
+    const SAFETY_INTERVAL: i64 = 1;
+    const TIME_DELAY_IN_SEC: i64 = 5;
+
+    impl StreamReader {
+        fn test_new(
+            session: &Arc<Session>,
+            stream_ids: Vec<StreamID>,
+            start_timestamp: chrono::Duration,
+            window_size: chrono::Duration,
+            safety_interval: chrono::Duration,
+            sleep_interval: time::Duration,
+        ) -> StreamReader {
+            StreamReader {
+                session: session.clone(),
+                stream_id_vec: stream_ids,
+                lower_timestamp: start_timestamp,
+                window_size,
+                safety_interval,
+                upper_timestamp: Default::default(),
+                sleep_interval,
+            }
+        }
+    }
+
+    async fn setup(session: &Arc<Session>) -> Result<StreamReader, Box<dyn Error>> {
+        create_test_db(session).await?;
+        populate_test_db(session).await?;
+        let stream_id = get_cdc_stream_id(session).await?;
+
+        let start_timestamp: chrono::Duration = chrono::Duration::seconds(chrono::Local::now().timestamp() - TIME_DELAY_IN_SEC);
+        let sleep_interval: time::Duration = time::Duration::from_secs(SLEEP_INTERVAL);
+        let window_size: chrono::Duration = chrono::Duration::seconds(WINDOW_SIZE);
+        let safety_interval: chrono::Duration = chrono::Duration::seconds(SAFETY_INTERVAL);
+
+        let reader = StreamReader::test_new(
+            session,
+            Vec::from([stream_id]),
+            start_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+        );
+
+        Ok(reader)
+    }
+
+    fn get_create_table_query() -> String {
+        format!("CREATE TABLE {}.{} (pk int, t int, v text, s text, PRIMARY KEY (pk, t)) WITH cdc = {{'enabled':true}};",
+                TEST_KEYSPACE,
+                TEST_TABLE
+        )
+    }
+
+    async fn create_test_db(session: &Session) -> Result<(), Box<dyn Error>> {
+        let mut create_keyspace_query = Query::new(format!(
+            "CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = {{'class': 'SimpleStrategy', 'replication_factor': 1}};",
+            TEST_KEYSPACE
+        ));
+        create_keyspace_query.set_consistency(Consistency::All);
+
+        session.query(create_keyspace_query, &[]).await?;
```

Comment:
You should wait for schema agreement as in [source](https://github.com/piodul/scylla-cdc-rust/pull/8#issuecomment-966628527)